gdal/frmts/zarr/zarrdataset.cpp

```diff
@@ -0,0 +1,393 @@
+/******************************************************************************
+ *
+ * Project:  Zarr Reader
+ * Purpose:  All code for Zarr Reader
+ * Author:   David Brochart, david.brochart@gmail.com
+ *
+ ******************************************************************************
+ * Copyright (c) 2021, David Brochart <david.brochart@gmail.com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include "cpl_port.h"
+#include "gdal_frmts.h"
+#include "gdal_pam.h"
+
+#include "xtensor-zarr/xzarr_hierarchy.hpp"
+#include "xtensor-zarr/xzarr_gdal_store.hpp"
+#include "xtensor-zarr/xzarr_file_system_store.hpp"
+#include "xtensor-io/xio_gzip.hpp"
+//#include "xtensor-io/xio_blosc.hpp"  // TODO
+
+#include <algorithm>
+
+CPL_CVSID("$Id$")
+
+namespace
+{
+    template <typename T>
+    void assign_chunk(void* pImage, xt::zarray& z, int nBlockYSize, int nBlockXSize,
+                      int nBlockYOff, int nBlockXOff,
+                      xt::dynamic_shape<std::size_t>& chunk_shape)
+    {
+        try
+        {
+            xt::xstrided_slice_vector sv(
+                {xt::range(chunk_shape[0] * nBlockYOff,
+                           chunk_shape[0] * (nBlockYOff + 1)),
+                 xt::range(chunk_shape[1] * nBlockXOff,
+                           chunk_shape[1] * (nBlockXOff + 1))});
+            xt::zarray chunk = xt::strided_view(z, sv);
+            xt::xarray<T> typed_chunk = chunk.get_array<T>();
+            for( int i = 0; i < nBlockYSize; i++ )
+            {
+                for( int j = 0; j < nBlockXSize; j++ )
+                {
+                    static_cast<T*>(pImage)[i * chunk_shape[1] + j] = typed_chunk(i, j);
+                }
+            }
+        }
+        catch (const std::runtime_error& e)
+        {
+            CPLError(CE_Failure, CPLE_AppDefined, "Error while processing Zarr array.\n");
```

Comment:
EOL character not needed for CPLError(), and adding actual error message might help:
```suggestion
            CPLError(CE_Failure, CPLE_AppDefined, "Error while processing Zarr array: %s", e.what());
```

Reply:
Thanks!